File: DSA/src/Questions/BinaryTree/SerializeDeserialize.java
package Questions.BinaryTree;

public class SerializeDeserialize {

    // Encodes a tree to a single string.
    public static String serialize(TreeNode root) {
        if (root == null)
            return "#";
        return root.val + " " + serialize(root.left) + " " + serialize(root.right);
    }

    // Decodes your encoded data to tree.
    public static TreeNode deserialize(String data) {
        String[] arr = data.split(" ");
        int[] index = {0}; // Reset the index before deserialization starts
        return helper(arr, index);
    }

    private static TreeNode helper(String[] arr, int[] index) {
        if (index[0] >= arr.length || arr[index[0]].equals("#")) {
            index[0]++; // Increment the index when encountering "#" or reaching the end of array
            return null;
        }

        TreeNode root = new TreeNode(Integer.valueOf(arr[index[0]]));
        index[0]++; // Move to the next element in the array
        root.left = helper(arr,index);
        root.right = helper(arr, index);
        return root;
    }

    public static void print(TreeNode root) {
        if(root == null){
            System.out.println("#");
            return;
        }
        System.out.println(root.val);
        print(root.left);
        print(root.right);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(5);

        String serialized = serialize(root);
        System.out.println(serialized);
        TreeNode deserialized = deserialize(serialized);
        print(deserialized);

    }
}

File: DSA/Python/Questions/BinaryTree/SerializeDeserialize.py
"""
# Question: Serialize and Deserialize Binary Tree
# Link: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/

# Time Complexity: O(N)
# Space Complexity: O(N)

# Algorithm:
# 1. Preorder traversal for serialization
# 2. String parsing for deserialization
# 3. Null marker handling
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class Codec:
    def serialize(self, root: TreeNode) -> str:
        if not root:
            return "null"
        return f"{root.val},{self.serialize(root.left)},{self.serialize(root.right)}"

    def deserialize(self, data: str) -> TreeNode:
        def dfs() -> TreeNode:
            val = next(values)
            if val == "null":
                return None

            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node

        values = iter(data.split(","))
        return dfs()


def main():
    codec = Codec()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    serialized = codec.serialize(root)
    print(serialized)
    deserialized = codec.deserialize(serialized)
    print(codec.serialize(deserialized))


if __name__ == "__main__":
    main()
